File: server/src/main/kotlin/org/bitlap/server/raft/BitlapServerEndpoint.kt
package org.bitlap.server.raft

import com.alipay.sofa.jraft.Node
import com.alipay.sofa.jraft.RaftGroupService
import com.alipay.sofa.jraft.conf.Configuration
import com.alipay.sofa.jraft.entity.PeerId
import com.alipay.sofa.jraft.option.NodeOptions
import com.alipay.sofa.jraft.rpc.RaftRpcServerFactory
import com.alipay.sofa.jraft.rpc.RpcServer
import com.alipay.sofa.jraft.rpc.impl.MarshallerHelper
import com.alipay.sofa.jraft.util.RpcFactoryHelper
import org.apache.commons.io.FileUtils
import org.bitlap.common.BitlapConf
import org.bitlap.common.LifeCycleWrapper
import org.bitlap.common.utils.withPaths
import org.bitlap.network.NetworkHelper
import org.bitlap.network.core.NetworkServiceImpl
import org.bitlap.network.core.SessionManager
import org.bitlap.network.processor.CloseSessionProcessor
import org.bitlap.network.processor.ExecuteStatementProcessor
import org.bitlap.network.processor.FetchResultsProcessor
import org.bitlap.network.processor.GetColumnsProcessor
import org.bitlap.network.processor.GetResultSetMetaDataProcessor
import org.bitlap.network.processor.GetSchemasProcessor
import org.bitlap.network.processor.GetTablesProcessor
import org.bitlap.network.processor.OpenSessionProcessor
import java.io.File

/**
 * Desc: Endpoint of bitlap server
 *
 * Mail: chk19940609@gmail.com
 * Created by IceMimosa
 * Date: 2021/4/22
 */
open class BitlapServerEndpoint(private val conf: BitlapConf) : LifeCycleWrapper(), NetworkHelper {

    private lateinit var node: Node

    @Synchronized
    override fun start() {
        if (this.started) {
            return
        }
        super.start()
        val serverIdStr = conf.get(BitlapConf.NODE_BIND_HOST)
        val nodeOptions = extractOptions(conf)
        val serverId = PeerId().apply {
            require(parse(serverIdStr)) { "Fail to parse serverId:$serverIdStr" }
        }
        registerMessageInstances(NetworkHelper.requestInstances()) {
            RpcFactoryHelper.rpcFactory().registerProtobufSerializer(it.first, it.second)
        }
        registerMessageInstances(NetworkHelper.responseInstances()) {
            MarshallerHelper.registerRespInstance(it.first, it.second)
        }
        val rpcServer = RaftRpcServerFactory.createRaftRpcServer(serverId.endpoint)
        registerProcessor(rpcServer)
        val raftGroupService = RaftGroupService(conf.get(BitlapConf.NODE_GROUP_ID), serverId, nodeOptions, rpcServer)
        this.node = raftGroupService.start()
        println("Started counter server at port:" + node.nodeId.peerId.port)
    }

    @Synchronized
    override fun close() {
        super.close()
        this.node.shutdown()
    }
}

private fun registerProcessor(rpcServer: RpcServer) {
    val cliService = NetworkServiceImpl(SessionManager())
    listOf(
        CloseSessionProcessor(cliService),
        OpenSessionProcessor(cliService),
        ExecuteStatementProcessor(cliService),
        FetchResultsProcessor(cliService),
        GetResultSetMetaDataProcessor(cliService),
        GetSchemasProcessor(cliService),
        GetTablesProcessor(cliService),
        GetColumnsProcessor(cliService),
    ).forEach { rpcServer.registerProcessor(it) }
}

private fun extractOptions(conf: BitlapConf): NodeOptions {
    val dataPath = conf.get(BitlapConf.DEFAULT_ROOT_DIR_LOCAL)!!
    val initConfStr = conf.get(BitlapConf.NODE_BIND_PEERS)
    val raftTimeout: Int = conf.get(BitlapConf.NODE_RAFT_TIMEOUT).let { if (it.isNullOrEmpty()) 1 else it.toInt() } * 1000
    FileUtils.forceMkdir(File(dataPath))
    return NodeOptions().apply {
        logUri = dataPath.withPaths("raft", "log")
        raftMetaUri = dataPath.withPaths("raft", "meta")
        snapshotUri = dataPath.withPaths("raft", "snapshot")
        FileUtils.forceMkdir(File(logUri))
        FileUtils.forceMkdir(File(raftMetaUri))
        FileUtils.forceMkdir(File(snapshotUri))
        electionTimeoutMs = raftTimeout
        isDisableCli = false
        snapshotIntervalSecs = 30
        fsm = MetaStateMachine()
        initialConf = Configuration().apply {
            require(parse(initConfStr)) { "Fail to parse initConf: $initConfStr" }
        }
    }
}

File: server/src/main/scala/org/bitlap/server/raft/BitlapServerEndpoint.scala
package org.bitlap.server.raft

import com.alipay.sofa.jraft.conf.Configuration
import com.alipay.sofa.jraft.entity.PeerId
import com.alipay.sofa.jraft.option.NodeOptions
import com.alipay.sofa.jraft.rpc.impl.MarshallerHelper
import com.alipay.sofa.jraft.rpc.{RaftRpcServerFactory, RpcServer}
import com.alipay.sofa.jraft.util.RpcFactoryHelper
import com.alipay.sofa.jraft.{Node, RaftGroupService}
import org.apache.commons.io.FileUtils
import org.bitlap.common.utils.StringEx
import org.bitlap.common.{BitlapConf, LifeCycleThread}
import org.bitlap.net.processor._
import org.bitlap.net.session.SessionManager
import org.bitlap.net.{NetworkHelper, NetworkServiceImpl}

import java.io.File
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

/**
 *
 * @author 梦境迷离
 * @version 1.0,2021/12/3
 */
class BitlapServerEndpoint(private val conf: BitlapConf) extends LifeCycleThread("BitlapServerEndpoint") with NetworkHelper {

  private var node: Node = _

  override def run(): Unit = {
    val serverIdStr = conf.get(BitlapConf.NODE_BIND_HOST)
    val nodeOptions = extractOptions()
    val serverId = new PeerId()
    serverId.parse(serverIdStr)
    registerMessageInstances(NetworkHelper.requestInstances(), (t1, t2) =>
      RpcFactoryHelper.rpcFactory().registerProtobufSerializer(t1, t2)
    )
    registerMessageInstances(NetworkHelper.responseInstances(), { (t1, t2) =>
      MarshallerHelper.registerRespInstance(t1, t2)
    })
    val rpcServer = RaftRpcServerFactory.createRaftRpcServer(serverId.getEndpoint)
    registerProcessor(rpcServer)
    val raftGroupService = new RaftGroupService(conf.get(BitlapConf.NODE_GROUP_ID), serverId, nodeOptions, rpcServer)
    this.node = raftGroupService.start()
    println("Started counter server at port:" + node.getNodeId.getPeerId.getPort)
  }

  override def close(): Unit = {
    this.synchronized {
      super.close()
      this.node.shutdown()
    }
  }

  def extractOptions(): NodeOptions = {
    val dataPath = conf.get(BitlapConf.DEFAULT_ROOT_DIR_LOCAL)
    val initConfStr = conf.get(BitlapConf.NODE_BIND_PEERS)
    val raftTimeout = conf.get(BitlapConf.NODE_RAFT_TIMEOUT)
    FileUtils.forceMkdir(new File(dataPath))
    val logUri = StringEx.withPaths(dataPath, "raft", "log")
    val raftMetaUri = StringEx.withPaths(dataPath, "raft", "meta")
    val snapshotUri = StringEx.withPaths(dataPath, "raft", "snapshot")

    val nodeOptions = new NodeOptions()
    nodeOptions.setLogUri(logUri)
    nodeOptions.setRaftMetaUri(raftMetaUri)
    nodeOptions.setSnapshotUri(snapshotUri)
    nodeOptions.setFsm(new MetaStateMachine)
    nodeOptions.setElectionTimeoutMs(raftTimeout.toInt)
    nodeOptions.setDisableCli(false)

    val cc = new Configuration()
    cc.parse(initConfStr)
    nodeOptions.setInitialConf(cc)

    FileUtils.forceMkdir(new File(logUri))
    FileUtils.forceMkdir(new File(raftMetaUri))
    FileUtils.forceMkdir(new File(snapshotUri))
    nodeOptions
  }

  private def registerProcessor(rpcServer: RpcServer): Unit = {
    val sessionManager = new SessionManager()
    Future {
      sessionManager.startListener()
    }
    val cliService = new NetworkServiceImpl(sessionManager)
    val processors = ProcessorsManager(cliService, null)
    List(
      processors.closeSession,
      processors.openSession,
      processors.executeStatement,
      processors.fetchResults,
      processors.getResultSet,
      processors.getSchemas,
      processors.getColumns,
      processors.getTables
    ).foreach {
      rpcServer.registerProcessor(_)
    }
  }
}

